spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCdcSparkAction.java

```diff
@@ -0,0 +1,267 @@
+    Snapshot snapshot = table.snapshot(snapshotId);
+    ManifestGroup manifestGroup = new ManifestGroup(table.io(), snapshot.dataManifests(), snapshot.deleteManifests())
+        .filterData(filter)
+        .ignoreAdded()
+        .specsById(((HasTableOperations) table).operations().current().specsById());
+
+    return ImmutableList.copyOf(manifestGroup.planFiles());
+  }
+
+  private Dataset<Row> readRowLevelDeletes(long snapshotId, int commitOrder) {
+    List<FileScanTask> fileScanTasks = planRowLevelDeleteFiles(snapshotId);
+    if (fileScanTasks.isEmpty()) {
+      return null;
+    }
+
+    String groupID = UUID.randomUUID().toString();
+    FileScanTaskSetManager manager = FileScanTaskSetManager.get();
+    manager.stageTasks(table, groupID, fileScanTasks);
+    Dataset<Row> scanDF = spark().read().format("iceberg")
+        .option(SparkReadOptions.FILE_SCAN_TASK_SET_ID, groupID)
+        .load(table.name())
+        .filter(functions.column(MetadataColumns.IS_DELETED.name()).equalTo(true));
+
+    return withCdcColumns(scanDF, snapshotId, "D", commitOrder);
+  }
+
+  private List<FileScanTask> planRowLevelDeleteFiles(long snapshotId) {
+    Snapshot snapshot = table.snapshot(snapshotId);
+    List<ManifestFile> manifestFiles = snapshot.deleteManifests().stream()
+        .filter(manifestFile -> manifestFile.snapshotId().equals(snapshotId))
+        .collect(Collectors.toList());
+
+    // todo create a partition filter to filter out unrelated partitions
+    ManifestGroup manifestGroup = new ManifestGroup(table.io(), snapshot.dataManifests(), manifestFiles)
+        .filterData(filter)
+        .onlyWithDeletes()
+        .specsById(((HasTableOperations) table).operations().current().specsById());
+
+    return ImmutableList.copyOf(manifestGroup.planFiles());
+  }
+
+  private Dataset<Row> withCdcColumns(Dataset<Row> df, long snapshotId, String cdcType, int commitOrder) {
+    return df.withColumn(RECORD_TYPE, lit(cdcType))
+        .withColumn(COMMIT_SNAPSHOT_ID, lit(snapshotId))
+        .withColumn(COMMIT_TIMESTAMP, lit(table.snapshot(snapshotId).timestampMillis()))
+        .withColumn(COMMIT_ORDER, lit(commitOrder));
+  }
+
+  @Override
+  public Cdc useSnapshot(long snapshotId) {
+    if (table.snapshot(snapshotId) != null) {
+      snapshotIds.clear();
+      snapshotIds.add(snapshotId);
+    }
+    return this;
+  }
+
+  @Override
+  public Cdc between(long fromSnapshotId, long toSnapshotId) {
+    Preconditions.checkArgument(table.snapshot(fromSnapshotId) != null,
+        "The fromSnapshotId(%s) is invalid", fromSnapshotId);
+    Preconditions.checkArgument(table.snapshot(toSnapshotId) != null,
+        "The toSnapshotId(%s) is invalid", toSnapshotId);
+    Preconditions.checkArgument(SnapshotUtil.isAncestorOf(table, toSnapshotId, fromSnapshotId),
+        "The fromSnapshot(%s) is not an ancestor of the toSnapshot(%s)", fromSnapshotId, toSnapshotId);
+
+    snapshotIds.clear();
+    // include the fromSnapshotId
```

Comment:
exclusive behavior for `fromSnapshotId` might be easier to work with. Assuming last enumerated position (snapshotId) needs to be saved somewhere. Next run would scan the range of `(fromSnapshotId, toSnapshotId]`

Reply:
Make sense. One concern is that it cannot get the first snapshot of a table. Two solutions:
1. Add a new interface like `toSnapshot()`, which compute all snapshots to the `toSnapshot`. 
2. Add a overwrite method `between(long fromSnapshotId, long toSnapshotId, boolean includeFromSnapshotId)`
What do you think?

Reply:
what do you mean "get the first snapshot of a table"?

> Add a new interface like toSnapshot(), which compute all snapshots to the toSnapshot.

This is covered if `fromSnapshotId` is not set, which means null is used. Then `IncrementalScan` would process all snapshots up to the `toSnapshotId`. As if we trace the ancestor chain of `toSnapshotId`, we will eventually trace to `null` snapshotId and stop. 

>Add a overwrite method between(long fromSnapshotId, long toSnapshotId, boolean includeFromSnapshotId)

This API is not not necessary. we can just set the `fromSnapshotId` to the parent snapshot id if we want the inclusive behavior.

Reply:
`null` can solve the issue. However, I was trying to avoid it since it doesn't have a clear meaning to users. 
1. Users don't know they can input a null.
2. Users don't know what will happen when they input a null.

So users have to rely on the doc to figure that out. I'd suggest to provide the new interface `beforeSnapshot()` or `toSnapshot()`. It did the same thing under the hood, but more meaningful to users. Another benefit is that it'd be logically complete since @aokolnychyi also proposed the method `afterSnapshot()` in here, https://github.com/apache/iceberg/pull/4539#discussion_r855718325.

Reply:
yeah. @rdblue also made a similar suggestion: https://github.com/apache/iceberg/pull/4580#discussion_r856412067. We will go with `fromSnapshotInclusive` and `fromSnapshotExclusive` in the new increment scan interface